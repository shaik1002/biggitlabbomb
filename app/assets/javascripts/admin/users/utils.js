import { queryToObject } from '~/lib/utils/url_utility';
import { TOKENS } from './constants';

export const generateUserPaths = (paths, id) => {
  return Object.fromEntries(
    Object.entries(paths).map(([action, genericPath]) => {
      return [action, genericPath.replace('id', id)];
    }),
  );
};

/**
 * @typedef {{type: string, value: {data: string, operator: string}}} Token
 */

/**
 * Initialize token values based on the URL parameters
 * @param {string} query - document.location.searchd
 *
 * @returns {{tokens: Array<string|Token>, sort: string}}
 */
export function initializeValuesFromQuery(query = document.location.search) {
  const tokens = [];

  const { filter, search_query: searchQuery, sort } = queryToObject(query);

  if (filter) {
    const token = TOKENS.find(({ options }) => options.some(({ value }) => value === filter));

    if (token) {
      tokens.push({
        type: token.type,
        value: {
          data: filter,
          operator: token.operators[0].value,
        },
      });
    }
  }

  if (searchQuery) {
    tokens.push(searchQuery);
  }

  return { tokens, sort };
}
