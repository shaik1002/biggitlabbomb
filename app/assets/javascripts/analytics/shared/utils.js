import { flatten } from 'lodash';
import dateFormat from '~/lib/dateformat';
import { slugify } from '~/lib/utils/text_utility';
import { joinPaths } from '~/lib/utils/url_utility';
import { urlQueryToFilter } from '~/vue_shared/components/filtered_search_bar/filtered_search_utils';
import { dateFormats, METRICS_POPOVER_CONTENT } from './constants';

export const filterBySearchTerm = (data = [], searchTerm = '', filterByKey = 'name') => {
  if (!searchTerm?.length) return data;
  return data.filter((item) => item[filterByKey].toLowerCase().includes(searchTerm.toLowerCase()));
};

export const toYmd = (date) => dateFormat(date, dateFormats.isoDate);

/**
 * Takes a url and extracts query parameters used for the shared
 * filter bar
 *
 * @param {string} url The URL to extract query parameters from
 * @returns {Object}
 */
export const extractFilterQueryParameters = (url = '') => {
  const {
    source_branch_name: selectedSourceBranch = null,
    target_branch_name: selectedTargetBranch = null,
    author_username: selectedAuthor = null,
    milestone_title: selectedMilestone = null,
    assignee_username: selectedAssigneeList = [],
    label_name: selectedLabelList = [],
  } = urlQueryToFilter(url);

  return {
    selectedSourceBranch,
    selectedTargetBranch,
    selectedAuthor,
    selectedMilestone,
    selectedAssigneeList,
    selectedLabelList,
  };
};

/**
 * Takes a url and extracts sorting and pagination query parameters into an object
 *
 * @param {string} url The URL to extract query parameters from
 * @returns {Object}
 */
export const extractPaginationQueryParameters = (url = '') => {
  const { sort, direction, page } = urlQueryToFilter(url);
  return {
    sort: sort?.value || null,
    direction: direction?.value || null,
    page: page?.value || null,
  };
};

export const getDataZoomOption = ({
  totalItems = 0,
  maxItemsPerPage = 40,
  dataZoom = [{ type: 'slider', bottom: 10, start: 0 }],
}) => {
  if (totalItems <= maxItemsPerPage) {
    return {};
  }

  const intervalEnd = Math.ceil((maxItemsPerPage / totalItems) * 100);

  return dataZoom.map((item) => {
    return {
      ...item,
      end: intervalEnd,
    };
  });
};

export const removeFlash = (type = 'alert') => {
  // flash-warning don't have dismiss button.
  document.querySelector(`.flash-${type} .js-close`)?.click();
};

/**
 * Prepares metric data to be rendered in the metric_card component
 *
 * @param {MetricData[]} data - The metric data to be rendered
 * @param {Object} popoverContent - Key value pair of data to display in the popover
 * @returns {TransformedMetricData[]} An array of metrics ready to render in the metric_card
 */
export const prepareTimeMetricsData = (data = [], popoverContent = {}) =>
  data.map(({ title: label, identifier, ...rest }) => {
    const metricIdentifier = identifier || slugify(label);
    return {
      ...rest,
      label,
      identifier: metricIdentifier,
      description: popoverContent[metricIdentifier]?.description || '',
    };
  });

const requestData = ({ request, endpoint, requestPath, params, name }) => {
  return request({ endpoint, params, requestPath })
    .then(({ data }) => data)
    .catch(() => {
      throw new Error(name);
    });
};

/**
 * Takes a configuration array of metrics requests (key metrics and DORA) and returns
 * a flat array of all the responses. Different metrics are retrieved from different endpoints
 * additionally we only support certain metrics for FOSS users.
 *
 * @param {Array} requests - array of metric api requests to be made
 * @param {String} requestPath - path for the group / project we are requesting
 * @param {Object} params - optional parameters to filter, including `created_after` and `created_before` dates
 * @returns a flat array of metrics
 */
export const fetchMetricsData = (requests = [], requestPath, params) => {
  const promises = requests.map((r) => requestData({ ...r, requestPath, params }));
  return Promise.all(promises).then((responses) =>
    prepareTimeMetricsData(flatten(responses), METRICS_POPOVER_CONTENT),
  );
};

/**
 * Generates a URL link to the VSD dashboard based on the group
 * and project paths passed into the method.
 *
 * @param {String} groupPath - Path of the specified group
 * @param {Array} projectPaths - Array of project paths to include in the `query` parameter
 * @returns a URL or blank string if there is no groupPath set
 */
export const generateValueStreamsDashboardLink = (namespacePath, projectPaths = []) => {
  if (namespacePath.length) {
    const query = projectPaths.length ? `?query=${projectPaths.join(',')}` : '';
    const dashboardsSlug = '/-/analytics/dashboards/value_streams_dashboard';
    const segments = [gon.relative_url_root || '', '/', namespacePath, dashboardsSlug];
    return joinPaths(...segments).concat(query);
  }
  return '';
};

/**
 * Extracts the relevant feature and license flags needed for VSA
 *
 * @param {Object} gon the global `window.gon` object populated when the page loads
 * @returns an object containing the extracted feature flags and their boolean status
 */
export const extractVSAFeaturesFromGON = () => ({
  // licensed feature toggles
  cycleAnalyticsForGroups: Boolean(gon?.licensed_features?.cycleAnalyticsForGroups),
  cycleAnalyticsForProjects: Boolean(gon?.licensed_features?.cycleAnalyticsForProjects),
  groupLevelAnalyticsDashboard: Boolean(gon?.licensed_features?.groupLevelAnalyticsDashboard),
  // feature flags
  vsaGroupAndProjectParity: Boolean(gon?.features?.vsaGroupAndProjectParity),
});
