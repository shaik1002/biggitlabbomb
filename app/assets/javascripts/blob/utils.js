import { getBaseURL } from '~/lib/utils/url_utility';

const blameLinesPerPage = document.querySelector('.js-per-page')?.dataset?.blamePerPage;

export const getPageParamValue = (lineNum, blamePerPage = blameLinesPerPage) => {
  if (!blamePerPage) return '';
  const page = Math.ceil(parseInt(lineNum, 10) / parseInt(blamePerPage, 10));
  return page <= 1 ? '' : page;
};

export const getPageSearchString = (path, page) => {
  if (!page) return '';
  const url = new URL(path, getBaseURL());
  url.searchParams.set('page', page);
  return url.search;
};

export default () => ({});
