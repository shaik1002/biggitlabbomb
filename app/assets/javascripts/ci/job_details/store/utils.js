import { parseBoolean } from '~/lib/utils/common_utils';

/**
 * Filters out lines that have an offset lower than the offset provided.
 *
 * If no offset is provided, all the lines are returned back.
 *
 * @param {Array} newLines
 * @param {Number} offset
 * @returns Lines to be added to the log that have not been added.
 */
const linesAfterOffset = (newLines = [], offset = -1) => {
  if (offset === -1) {
    return newLines;
  }
  return newLines.filter((newLine) => newLine.offset > offset);
};

/**
 * Parses a series of trace lines from a job and returns lines and
 * sections of the log. Each line is annotated with a lineNumber.
 *
 * Sections have a range: starting line and ending line, plus a
 * "duration" string.
 *
 * @param {Array} newLines - Lines to add to the log
 * @param {Object} currentState - Current log: lines and sections
 * @returns Consolidated lines and sections to be displayed
 */
export const logLinesParser = (
  newLines = [],
  { currentLines = [], currentSections = {} } = {},
  hash = '',
) => {
  const lastCurrentLine = currentLines[currentLines.length - 1];
  const newLinesToAppend = linesAfterOffset(newLines, lastCurrentLine?.offset);

  if (!newLinesToAppend.length) {
    return { lines: currentLines, sections: currentSections };
  }

  let lineNumber = lastCurrentLine?.lineNumber || 0;
  const lines = [...currentLines];
  const sections = { ...currentSections };

  newLinesToAppend.forEach((line) => {
    const {
      offset,
      content,
      section,
      section_header: isHeader,
      section_footer: isFooter,
      section_duration: duration,
      section_options: options,
    } = line;

    if (content.length) {
      lineNumber += 1;
      lines.push({
        offset,
        lineNumber,
        content,
        ...(section ? { section } : {}),
        ...(isHeader ? { isHeader: true } : {}),
      });
    }

    // root level lines have no section, skip creating one
    if (section) {
      sections[section] = sections[section] || {
        startLineNumber: 0,
        endLineNumber: Infinity, // by default, sections are unbounded / have no end
        duration: null,
        isClosed: false,
      };

      if (isHeader) {
        sections[section].startLineNumber = lineNumber;
      }
      if (options) {
        let isClosed = parseBoolean(options?.collapsed);
        // if a hash is present in the URL then we ensure
        // all sections are visible so we can scroll to the hash
        // in the DOM
        if (hash) {
          isClosed = false;
        }
        sections[section].isClosed = isClosed;

        const hideDuration = parseBoolean(options?.hide_duration);
        if (hideDuration) {
          sections[section].hideDuration = hideDuration;
        }
      }
      if (duration) {
        sections[section].duration = duration;
      }
      if (isFooter) {
        sections[section].endLineNumber = lineNumber;
      }
    }
  });

  return { lines, sections };
};
