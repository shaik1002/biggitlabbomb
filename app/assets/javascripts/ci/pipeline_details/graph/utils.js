import { isEmpty } from 'lodash';
import { getIdFromGraphQLId, etagQueryHeaders } from '~/graphql_shared/utils';
import { reportToSentry } from '~/ci/utils';

import { listByLayers } from '~/ci/pipeline_details/utils/parsing_utils';
import { unwrapStagesWithNeedsAndLookup } from '~/ci/pipeline_details/utils/unwrapping_utils';
import { beginPerfMeasure, finishPerfMeasureAndSend } from './perf_utils';

export { toggleQueryPollingByVisibility } from '~/graphql_shared/utils';

const addMulti = (mainPipelineProjectPath, linkedPipeline) => {
  return {
    ...linkedPipeline,
    multiproject: mainPipelineProjectPath !== linkedPipeline.project.fullPath,
  };
};

const calculatePipelineLayersInfo = (pipeline, componentName, metricsPath) => {
  const shouldCollectMetrics = Boolean(metricsPath);

  if (shouldCollectMetrics) {
    beginPerfMeasure();
  }

  let layers = null;

  try {
    layers = listByLayers(pipeline);

    if (shouldCollectMetrics) {
      finishPerfMeasureAndSend(layers.linksData.length, layers.numGroups, metricsPath);
    }
  } catch (err) {
    reportToSentry(componentName, err);
  }

  return layers;
};

const getQueryHeaders = (etagResource) =>
  etagQueryHeaders('verify/ci/pipeline-graph', etagResource);

const serializeGqlErr = (gqlError) => {
  const { locations = [], message = '', path = [] } = gqlError;

  // eslint-disable-next-line @gitlab/require-i18n-strings
  return `
    ${message}.
    Locations: ${locations
      .flatMap((loc) => Object.entries(loc))
      .flat(2)
      .join(' ')}.
    Path: ${path.join(', ')}.
  `;
};

const serializeLoadErrors = (errors) => {
  const { gqlError, graphQLErrors, networkError, message } = errors;

  if (!isEmpty(graphQLErrors)) {
    return graphQLErrors.map((err) => serializeGqlErr(err)).join('; ');
  }

  if (!isEmpty(gqlError)) {
    return serializeGqlErr(gqlError);
  }

  if (!isEmpty(networkError)) {
    return `Network error: ${networkError.message}`; // eslint-disable-line @gitlab/require-i18n-strings
  }

  return message;
};

const transformId = (linkedPipeline) => {
  return { ...linkedPipeline, id: getIdFromGraphQLId(linkedPipeline.id) };
};

const unwrapPipelineData = (mainPipelineProjectPath, data) => {
  if (!data?.project?.pipeline) {
    return null;
  }

  const { pipeline } = data.project;

  const {
    upstream,
    downstream,
    stages: { nodes: stages },
  } = pipeline;

  const { stages: updatedStages, lookup } = unwrapStagesWithNeedsAndLookup(stages);

  return {
    ...pipeline,
    id: getIdFromGraphQLId(pipeline.id),
    stages: updatedStages,
    stagesLookup: lookup,
    upstream: upstream
      ? [upstream].map(addMulti.bind(null, mainPipelineProjectPath)).map(transformId)
      : [],
    downstream: downstream
      ? downstream.nodes.map(addMulti.bind(null, mainPipelineProjectPath)).map(transformId)
      : [],
  };
};

const validateConfigPaths = (value) => value.graphqlResourceEtag?.length > 0;

export {
  calculatePipelineLayersInfo,
  getQueryHeaders,
  serializeGqlErr,
  serializeLoadErrors,
  unwrapPipelineData,
  validateConfigPaths,
};
