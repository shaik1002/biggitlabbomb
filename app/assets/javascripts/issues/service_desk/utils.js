import {
  OPERATOR_OR,
  TOKEN_TYPE_LABEL,
} from '~/vue_shared/components/filtered_search_bar/constants';
import { isSpecialFilter, isNotEmptySearchToken } from '~/issues/list/utils';
import {
  ALTERNATIVE_FILTER,
  NORMAL_FILTER,
  SPECIAL_FILTER,
  URL_PARAM,
} from '~/issues/list/constants';
import { filtersMap } from './constants';

const getFilterType = ({ type, value: { data, operator } }) => {
  const isUnionedLabel = type === TOKEN_TYPE_LABEL && operator === OPERATOR_OR;

  if (isUnionedLabel) {
    return ALTERNATIVE_FILTER;
  }
  if (isSpecialFilter(type, data)) {
    return SPECIAL_FILTER;
  }
  return NORMAL_FILTER;
};

export const convertToUrlParams = (filterTokens) => {
  const urlParamsMap = filterTokens.filter(isNotEmptySearchToken).reduce((acc, token) => {
    const filterType = getFilterType(token);
    const urlParam = filtersMap[token.type][URL_PARAM][token.value.operator]?.[filterType];
    return acc.set(
      urlParam,
      acc.has(urlParam) ? [acc.get(urlParam), token.value.data].flat() : token.value.data,
    );
  }, new Map());

  return Object.fromEntries(urlParamsMap);
};
