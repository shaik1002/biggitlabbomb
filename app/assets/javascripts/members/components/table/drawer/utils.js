import axios from 'axios';
import { roleDropdownItems } from '~/members/utils';
import {
  GROUP_LINK_ACCESS_LEVEL_PROPERTY_NAME,
  MEMBER_ACCESS_LEVEL_PROPERTY_NAME,
  MEMBERS_TAB_TYPES,
} from '~/members/constants';

export const getMemberRole = (roles, member) => {
  const { stringValue, integerValue, memberRoleId = null } = member.accessLevel;
  const role = roles.find(({ accessLevel }) => accessLevel === member.accessLevel.integerValue);

  return role || { text: stringValue, value: integerValue, memberRoleId };
};

// EE version has a special implementation, CE version just returns the basic version.
export const getRoleDropdownItems = roleDropdownItems;

// The API to update members uses different property names for the access level, depending on if it's a user or a group.
// Users use 'access_level', groups use 'group_access'.
const ACCESS_LEVEL_PROPERTY_NAME = {
  [MEMBERS_TAB_TYPES.user]: MEMBER_ACCESS_LEVEL_PROPERTY_NAME,
  [MEMBERS_TAB_TYPES.group]: GROUP_LINK_ACCESS_LEVEL_PROPERTY_NAME,
};

export const callRoleUpdateApi = async (member, role) => {
  const accessLevelProp = ACCESS_LEVEL_PROPERTY_NAME[member.namespace];

  return axios.put(member.memberPath, {
    [accessLevelProp]: role.accessLevel,
    member_role_id: role.memberRoleId,
  });
};

export const setMemberRole = (member, role) => {
  // eslint-disable-next-line no-param-reassign
  member.accessLevel = {
    stringValue: role.text,
    integerValue: role.accessLevel,
    description: role.description,
    memberRoleId: role.memberRoleId,
  };
};
