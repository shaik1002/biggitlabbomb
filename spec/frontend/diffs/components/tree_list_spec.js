import Vue, { nextTick } from 'vue';
// eslint-disable-next-line no-restricted-imports
import Vuex from 'vuex';
import TreeList from '~/diffs/components/tree_list.vue';
import createStore from '~/diffs/store/modules';
import DiffFileRow from '~/diffs/components//diff_file_row.vue';
import { stubComponent } from 'helpers/stub_component';
import { shallowMountExtended } from 'helpers/vue_test_utils_helper';
import { SET_PINNED_FILE_HASH, SET_TREE_DATA, SET_DIFF_FILES } from '~/diffs/store/mutation_types';
import { generateTreeList } from '~/diffs/utils/tree_worker_utils';
import { sortTree } from '~/ide/stores/utils';

describe('Diffs tree list component', () => {
  let wrapper;
  let store;
  let setRenderTreeListMock;
  const getScroller = () => wrapper.findComponent({ name: 'RecycleScroller' });
  const getFileRow = () => wrapper.findComponent(DiffFileRow);
  const findDiffTreeSearch = () => wrapper.findByTestId('diff-tree-search');

  Vue.use(Vuex);

  const createComponent = ({ hideFileStats = false } = {}) => {
    wrapper = shallowMountExtended(TreeList, {
      store,
      propsData: { hideFileStats },
      stubs: {
        // eslint will fail if we import the real component
        RecycleScroller: stubComponent(
          {
            name: 'RecycleScroller',
            props: {
              items: null,
            },
          },
          {
            template: '<div><slot :item="{ tree: [] }"></slot></div>',
          },
        ),
      },
    });
  };

  beforeEach(() => {
    const { getters, mutations, actions, state } = createStore();

    setRenderTreeListMock = jest.fn();

    store = new Vuex.Store({
      modules: {
        diffs: {
          namespaced: true,
          state: {
            isTreeLoaded: true,
            diffFiles: ['test'],
            addedLines: 10,
            removedLines: 20,
            mergeRequestDiff: {},
            realSize: 20,
            ...state,
          },
          getters: {
            allBlobs: getters.allBlobs,
            flatBlobsList: getters.flatBlobsList,
            pinnedFile: getters.pinnedFile,
          },
          mutations: { ...mutations },
          actions: {
            toggleTreeOpen: actions.toggleTreeOpen,
            goToFile: actions.goToFile,
            setRenderTreeList: setRenderTreeListMock,
          },
        },
      },
    });
  });

  const setupFilesInState = () => {
    const treeEntries = {
      app: {
        key: 'app',
        path: 'app',
        name: 'app',
        type: 'tree',
        tree: [],
        opened: true,
      },
      'index.js': {
        addedLines: 0,
        changed: true,
        deleted: false,
        fileHash: 'test',
        key: 'index.js',
        name: 'index.js',
        path: 'app/index.js',
        removedLines: 0,
        tempFile: true,
        type: 'blob',
        parentPath: 'app',
        tree: [],
        file_path: 'app/index.js',
        file_hash: 'app-index',
      },
      'test.rb': {
        addedLines: 0,
        changed: true,
        deleted: false,
        fileHash: 'test',
        key: 'test.rb',
        name: 'test.rb',
        path: 'app/test.rb',
        removedLines: 0,
        tempFile: true,
        type: 'blob',
        parentPath: 'app',
        tree: [],
        file_path: 'app/test.rb',
        file_hash: 'app-test',
      },
      LICENSE: {
        addedLines: 0,
        changed: true,
        deleted: false,
        fileHash: 'LICENSE',
        key: 'LICENSE',
        name: 'LICENSE',
        path: 'LICENSE',
        removedLines: 0,
        tempFile: true,
        type: 'blob',
        parentPath: '/',
        tree: [],
        file_path: 'LICENSE',
        file_hash: 'LICENSE',
      },
    };

    Object.assign(store.state.diffs, {
      treeEntries,
      tree: [
        treeEntries.LICENSE,
        {
          ...treeEntries.app,
          tree: [treeEntries['index.js'], treeEntries['test.rb']],
        },
      ],
    });

    return treeEntries;
  };

  describe('default', () => {
    beforeEach(() => {
      createComponent();
    });

    it('renders empty text', () => {
      expect(wrapper.text()).toContain('No files found');
    });
  });

  it('renders file count', () => {
    createComponent();

    expect(wrapper.findByTestId('file-count').text()).toBe('20');
  });

  describe('with files', () => {
    beforeEach(() => {
      setupFilesInState();
      createComponent();
    });

    describe('search by file extension', () => {
      it('hides scroller for no matches', async () => {
        const input = findDiffTreeSearch();

        input.vm.$emit('input', '*.md');

        await nextTick();

        expect(getScroller().exists()).toBe(false);
        expect(wrapper.text()).toContain('No files found');
      });

      it.each`
        extension       | itemSize
        ${'*.js'}       | ${2}
        ${'index.js'}   | ${2}
        ${'app/*.js'}   | ${2}
        ${'*.js, *.rb'} | ${3}
      `('returns $itemSize item for $extension', async ({ extension, itemSize }) => {
        const input = findDiffTreeSearch();

        input.vm.$emit('input', extension);

        await nextTick();

        expect(getScroller().props('items')).toHaveLength(itemSize);
      });
    });

    it('renders tree', () => {
      expect(getScroller().props('items')).toHaveLength(4);
    });

    it('hides file stats', () => {
      createComponent({ hideFileStats: true });
      expect(getFileRow().props('hideFileStats')).toBe(true);
    });

    it('calls toggleTreeOpen when clicking folder', () => {
      jest.spyOn(store, 'dispatch').mockReturnValue(undefined);

      getFileRow().vm.$emit('toggleTreeOpen', 'app');

      expect(store.dispatch).toHaveBeenCalledWith('diffs/toggleTreeOpen', 'app');
    });

    it('renders when renderTreeList is false', async () => {
      store.state.diffs.renderTreeList = false;

      await nextTick();
      expect(getScroller().props('items')).toHaveLength(5);
    });
  });

  describe('with viewedDiffFileIds', () => {
    const viewedDiffFileIds = { fileId: '#12345' };

    beforeEach(() => {
      setupFilesInState();
      store.state.diffs.viewedDiffFileIds = viewedDiffFileIds;
    });

    it('passes the viewedDiffFileIds to the FileTree', async () => {
      createComponent();

      await nextTick();
      expect(getFileRow().props('viewedFiles')).toBe(viewedDiffFileIds);
    });
  });

  describe('pinned file', () => {
    const filePaths = [
      ['nested-1.rb', 'folder/sub-folder/'],
      ['nested-2.rb', 'folder/sub-folder/'],
      ['nested-3.rb', 'folder/sub-folder/'],
      ['1.rb', 'folder/'],
      ['2.rb', 'folder/'],
      ['3.rb', 'folder/'],
      ['single.rb', 'folder-single/'],
      ['root-first.rb'],
      ['root-last.rb'],
    ];

    const pinFile = (fileHash) => {
      store.commit(`diffs/${SET_PINNED_FILE_HASH}`, fileHash);
    };

    const setupFiles = (diffFiles) => {
      const { treeEntries, tree } = generateTreeList(diffFiles);
      store.commit(`diffs/${SET_DIFF_FILES}`, diffFiles);
      store.commit(`diffs/${SET_TREE_DATA}`, {
        treeEntries,
        tree: sortTree(tree),
      });
    };

    const createFile = (name, path = '') => ({
      file_hash: name,
      path: `${path}${name}`,
      new_path: `${path}${name}`,
      file_path: `${path}${name}`,
    });

    beforeEach(() => {
      createComponent();
      setupFiles(filePaths.map(([name, path]) => createFile(name, path)));
    });

    describe('files in folders', () => {
      it.each(filePaths.map((path) => path[0]))('pins %s file', async (pinnedFile) => {
        pinFile(pinnedFile);
        await nextTick();
        const items = getScroller().props('items');
        expect(
          items.map(
            (item) =>
              `${'─'.repeat(item.level * 2)}${item.type === 'tree' ? '📁' : ''}${
                item.name || item.path
              }`,
          ),
        ).toMatchSnapshot();
      });
    });
  });

  describe('tree view buttons', () => {
    it.each`
      toggle                | renderTreeList
      ${'list-view-toggle'} | ${false}
      ${'tree-view-toggle'} | ${true}
    `(
      'calls setRenderTreeListMock with `$renderTreeList` when clicking $toggle clicked',
      ({ toggle, renderTreeList }) => {
        createComponent();

        wrapper.findByTestId(toggle).vm.$emit('click');

        expect(setRenderTreeListMock).toHaveBeenCalledWith(expect.anything(), {
          renderTreeList,
        });
      },
    );

    it.each`
      selectedToggle        | deselectedToggle      | renderTreeList
      ${'list-view-toggle'} | ${'tree-view-toggle'} | ${false}
      ${'tree-view-toggle'} | ${'list-view-toggle'} | ${true}
    `(
      'sets $selectedToggle as selected when renderTreeList is $renderTreeList',
      ({ selectedToggle, deselectedToggle, renderTreeList }) => {
        store.state.diffs.renderTreeList = renderTreeList;

        createComponent();

        expect(wrapper.findByTestId(deselectedToggle).props('selected')).toBe(false);
        expect(wrapper.findByTestId(selectedToggle).props('selected')).toBe(true);
      },
    );
  });
});
