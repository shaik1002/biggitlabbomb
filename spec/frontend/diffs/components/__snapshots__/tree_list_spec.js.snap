// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Diffs tree list component pinned file files in folders pins 1.rb file 1`] = `
Array [
  "📁folder/",
  "──1.rb",
  "📁folder",
  "──📁sub-folder",
  "────nested-1.rb",
  "────nested-2.rb",
  "────nested-3.rb",
  "──2.rb",
  "──3.rb",
  "📁folder-single",
  "──single.rb",
  "root-first.rb",
  "root-last.rb",
]
`;

exports[`Diffs tree list component pinned file files in folders pins 2.rb file 1`] = `
Array [
  "📁folder/",
  "──2.rb",
  "📁folder",
  "──📁sub-folder",
  "────nested-1.rb",
  "────nested-2.rb",
  "────nested-3.rb",
  "──1.rb",
  "──3.rb",
  "📁folder-single",
  "──single.rb",
  "root-first.rb",
  "root-last.rb",
]
`;

exports[`Diffs tree list component pinned file files in folders pins 3.rb file 1`] = `
Array [
  "📁folder/",
  "──3.rb",
  "📁folder",
  "──📁sub-folder",
  "────nested-1.rb",
  "────nested-2.rb",
  "────nested-3.rb",
  "──1.rb",
  "──2.rb",
  "📁folder-single",
  "──single.rb",
  "root-first.rb",
  "root-last.rb",
]
`;

exports[`Diffs tree list component pinned file files in folders pins nested-1.rb file 1`] = `
Array [
  "📁folder/sub-folder/",
  "──nested-1.rb",
  "📁folder",
  "──📁sub-folder",
  "────nested-2.rb",
  "────nested-3.rb",
  "──1.rb",
  "──2.rb",
  "──3.rb",
  "📁folder-single",
  "──single.rb",
  "root-first.rb",
  "root-last.rb",
]
`;

exports[`Diffs tree list component pinned file files in folders pins nested-2.rb file 1`] = `
Array [
  "📁folder/sub-folder/",
  "──nested-2.rb",
  "📁folder",
  "──📁sub-folder",
  "────nested-1.rb",
  "────nested-3.rb",
  "──1.rb",
  "──2.rb",
  "──3.rb",
  "📁folder-single",
  "──single.rb",
  "root-first.rb",
  "root-last.rb",
]
`;

exports[`Diffs tree list component pinned file files in folders pins nested-3.rb file 1`] = `
Array [
  "📁folder/sub-folder/",
  "──nested-3.rb",
  "📁folder",
  "──📁sub-folder",
  "────nested-1.rb",
  "────nested-2.rb",
  "──1.rb",
  "──2.rb",
  "──3.rb",
  "📁folder-single",
  "──single.rb",
  "root-first.rb",
  "root-last.rb",
]
`;

exports[`Diffs tree list component pinned file files in folders pins root-first.rb file 1`] = `
Array [
  "root-first.rb",
  "📁folder",
  "──📁sub-folder",
  "────nested-1.rb",
  "────nested-2.rb",
  "────nested-3.rb",
  "──1.rb",
  "──2.rb",
  "──3.rb",
  "📁folder-single",
  "──single.rb",
  "root-last.rb",
]
`;

exports[`Diffs tree list component pinned file files in folders pins root-last.rb file 1`] = `
Array [
  "root-last.rb",
  "📁folder",
  "──📁sub-folder",
  "────nested-1.rb",
  "────nested-2.rb",
  "────nested-3.rb",
  "──1.rb",
  "──2.rb",
  "──3.rb",
  "📁folder-single",
  "──single.rb",
  "root-first.rb",
]
`;

exports[`Diffs tree list component pinned file files in folders pins single.rb file 1`] = `
Array [
  "📁folder-single/",
  "──single.rb",
  "📁folder",
  "──📁sub-folder",
  "────nested-1.rb",
  "────nested-2.rb",
  "────nested-3.rb",
  "──1.rb",
  "──2.rb",
  "──3.rb",
  "root-first.rb",
  "root-last.rb",
]
`;
